File: kotlin.web.demo.server/static/js/Completion.js
/*
 * Copyright 2000-2012 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Created with IntelliJ IDEA.
 * User: Natalia.Ukhorskaya
 * Date: 3/29/12
 * Time: 1:56 PM
 */

var CompletionFromServer = (function () {

    function CompletionFromServer() {

        var instance = {
            getCompletion: function (confType, project, filename, cursorLine, cursorCh) {
                var confTypeString = Configuration.getStringFromType(confType);
                getCompletion(confTypeString, project, filename, cursorLine, cursorCh);
            },
            onLoadCompletion:function (completionObject) {
                editor.showCompletionResult(completionObject);
                statusBarView.html(ActionStatusMessages.get_completion_ok);
            },
            onFail: function (error) {
            consoleView.writeException(error);
            statusBarView.html(ActionStatusMessages.get_completion_fail);
            }
        };

        var isLoadingCompletion = false;

        function getCompletion(confTypeString, project, filename, cursorLine, cursorCh) {
            if (!isLoadingCompletion) {
                isLoadingCompletion = true;
                $.ajax({
                    url: generateAjaxUrl("complete", "&runConf=" + confTypeString),
                    context: document.body,
                    success: function (data) {
                        isLoadingCompletion = false;
                        if (checkDataForNull(data)) {
                            if (checkDataForException(data)) {
                                instance.onLoadCompletion(data);
                            } else {
                                instance.onFail(data);
                            }
                        } else {
                            instance.onFail("Incorrect data format.");
                        }
                    },
                    dataType: "json",
                    type: "POST",
                    data: JSON.stringify({ project: project, filename: filename, line : cursorLine, ch : cursorCh}),
                    timeout: 10000,
                    error: function (jqXHR, textStatus, errorThrown) {
                        isLoadingCompletion = false;
                        instance.onFail(textStatus + " : " + errorThrown);
                    }
                });
            }
        }

        return instance;
    }


    return CompletionFromServer;
})();

File: kotlin.web.demo.frontend/src/views/editor/Completion.kt
/*
 * Copyright 2000-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package views.editor

import org.w3c.dom.HTMLElement
import kotlin.browser.document
import html4k.*
import html4k.js.*
import html4k.dom.*
import utils.editor.CodeMirror
import utils.editor.Position

data class Hint(val from: Position, val to: Position, var list: Array<Completion>)

class Completion(private val proposal: CompletionProposal) {
    val text = proposal.text
    val displayText = proposal.displayText

    fun render(element: HTMLElement, self: dynamic, data: dynamic) = element.append {
        div {
            classes = setOf("icon", proposal.icon)
        }
        div {
            +displayText
            classes = setOf("name")
        }
        div {
            +proposal.tail
            classes = setOf("tail")
        }
    }

    fun hint(cm: CodeMirror, self: dynamic, data: dynamic){
        var cur = cm.getCursor()
        val token = cm.getTokenAt(cm.getCursor())
        val replacement: String = data.text

        val from = Position(cur.line, token.start)
        var to = Position(cur.line, token.end)

        if ((token.string == ".") || (token.string == " ") || (token.string == "(")) {
            //Insert string
            cm.replaceRange(replacement, to, to);
        } else {
            cm.replaceRange(replacement, from, to);
            if (data.text.endsWith('(')) {
                cm.replaceRange(")", Position(cur.line, token.start + replacement.length()));
                cm.execCommand("goCharLeft")
            }
        }
    };
}

public data class CompletionProposal(val icon: String, val text: String, val displayText: String, val tail: String)